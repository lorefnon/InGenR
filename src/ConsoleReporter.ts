import chalk, { Chalk } from "chalk"
import { isNumber } from "lodash"
import { WarningEntry } from "./warnings"

export interface Reporter {
  bufferWarning(
    filePath: string,
    lineIndex?: number,
    line?: string,
    warnings?: WarningEntry[]
  ): void
  reportAllWarnings(): void
}

export interface FileWarning {
  filePath: string
  lineIndex?: number
  line?: string
  warnings: WarningEntry[]
}

interface ConsoleReporterOptions {
  enableColor?: boolean
}

export const defaultReporterOptions: ConsoleReporterOptions = {
  enableColor: !!process.stdout.isTTY
}

export class ConsoleReporter implements Reporter {
  warnings: Map<string, FileWarning[]>
  console: Console
  chalk: Chalk

  constructor(private options = defaultReporterOptions) {
    this.warnings = new Map()
    this.console = console
    this.chalk = new chalk.constructor({
      enabled: !!options.enableColor
    })
  }

  bufferWarning(
    filePath: string,
    lineIndex?: number,
    line?: string,
    warnings: WarningEntry[] = []
  ) {
    const fileWarnings = this.warnings.get(filePath) || []
    fileWarnings.push({ filePath, lineIndex, line, warnings })
    this.warnings.set(filePath, fileWarnings)
  }

  reportFileWarnings(fileWarnings: FileWarning[]) {
    if (!fileWarnings) return
    for (const { lineIndex, line, warnings } of fileWarnings) {
      for (const entry of warnings) {
        let preSpacer = ""
        if (isNumber(lineIndex)) {
          const lineNumRepr = `L:${lineIndex + 1} : `
          this.console.log(this.chalk.gray(lineNumRepr) + this.chalk.blue(line!))
          preSpacer = "^"
          let marginLeft = lineNumRepr.length
          if (entry.index) {
            marginLeft += entry.index
          }
          for (let i = 0; i < marginLeft; i++) {
            preSpacer = ` ${preSpacer}`
          }
        }
        this.console.log(this.chalk.red(`${preSpacer} ${entry.message}`))
        if (entry.error) {
          this.console.error(entry.error)
        }
      }
    }
  }

  reportAllWarnings() {
    for (const [filePath, fileWarnings] of this.warnings) {
      const count = fileWarnings.reduce((sum, fileWarning) => sum + fileWarning.warnings.length, 0)
      this.console.log(this.chalk.blue(`[InGenR] ${filePath}: ${count} warnings`))
      this.reportFileWarnings(fileWarnings)
    }
  }
}
