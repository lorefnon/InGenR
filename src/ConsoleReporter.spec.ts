import { ConsoleReporter } from "./ConsoleReporter"

describe("ConsoleReporter", () => {
  it("works as expected", () => {
    const reporter = new ConsoleReporter({
      enableColor: false
    })
    const messages: any[] = []
    reporter.console = ["log", "debug", "info", "warn", "error"].reduce(
      (result: any, type: string) => {
        result[type] = (...args: any[]) => {
          messages.push({ type, args })
        }
        return result
      },
      {}
    ) as any
    reporter.bufferWarning("/tmp/foo.js", 10, "lorem ipsum dolor sit amet", [
      { message: "something failed" },
      { index: 10, message: "something else failed" }
    ])
    reporter.bufferWarning("/tmp/bar.js", undefined, undefined, [
      {
        message: "something failed",
        error: new Error("Something Failed")
      }
    ])
    expect(messages).toHaveLength(0)
    reporter.reportAllWarnings()
    expect(messages).toMatchSnapshot()
  })
})
