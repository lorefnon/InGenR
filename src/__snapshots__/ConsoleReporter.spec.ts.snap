// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsoleReporter works as expected 1`] = `
Array [
  Object {
    "args": Array [
      "[InGenR] /tmp/foo.js: 2 warnings",
    ],
    "type": "log",
  },
  Object {
    "args": Array [
      "L:11 : lorem ipsum dolor sit amet",
    ],
    "type": "log",
  },
  Object {
    "args": Array [
      "       ^ something failed",
    ],
    "type": "log",
  },
  Object {
    "args": Array [
      "L:11 : lorem ipsum dolor sit amet",
    ],
    "type": "log",
  },
  Object {
    "args": Array [
      "                 ^ something else failed",
    ],
    "type": "log",
  },
  Object {
    "args": Array [
      "[InGenR] /tmp/bar.js: 1 warnings",
    ],
    "type": "log",
  },
  Object {
    "args": Array [
      " something failed",
    ],
    "type": "log",
  },
  Object {
    "args": Array [
      [Error: Something Failed],
    ],
    "type": "error",
  },
]
`;
